File: webmvp/src/lib/languageTags.ts
/** Standard language tags for library filtering. Format: `lang:<language>`. */
export const LANGUAGE_TAGS = [
  { value: "lang:english", label: "English" },
  { value: "lang:malayalam", label: "Malayalam" },
  { value: "lang:hindi", label: "Hindi" },
  { value: "lang:tamil", label: "Tamil" },
  { value: "lang:marathi", label: "Marathi" },
  { value: "lang:telugu", label: "Telugu" },
] as const;

export type LanguageTagValue = (typeof LANGUAGE_TAGS)[number]["value"];

const LANGUAGE_TAG_PREFIX = "lang:";

export function isLanguageTag(tag: string): tag is LanguageTagValue {
  return tag.startsWith(LANGUAGE_TAG_PREFIX);
}

export function languageLabel(tag: string): string | null {
  const match = LANGUAGE_TAGS.find((entry) => entry.value === tag);
  return match?.label ?? null;
}

export function getLanguageTag(tags: string[]): LanguageTagValue | null {
  const match = tags.find(isLanguageTag);
  return match ?? null;
}

/** Replace any existing language tag with the selected one, or remove all language tags. */
export function setLanguageTag(tags: string[], languageTag: LanguageTagValue | null): string[] {
  const withoutLanguage = tags.filter((tag) => !isLanguageTag(tag));
  if (!languageTag) {
    return withoutLanguage;
  }
  return [...withoutLanguage, languageTag];
}

File: mobile/lib/domain/language_tags.dart
class LanguageTagOption {
  const LanguageTagOption({required this.value, required this.label});

  final String value;
  final String label;
}

const List<LanguageTagOption> languageTags = [
  LanguageTagOption(value: 'lang:english', label: 'English'),
  LanguageTagOption(value: 'lang:malayalam', label: 'Malayalam'),
  LanguageTagOption(value: 'lang:hindi', label: 'Hindi'),
  LanguageTagOption(value: 'lang:tamil', label: 'Tamil'),
  LanguageTagOption(value: 'lang:marathi', label: 'Marathi'),
  LanguageTagOption(value: 'lang:telugu', label: 'Telugu'),
];

const String languageTagPrefix = 'lang:';

bool isLanguageTag(String tag) => tag.startsWith(languageTagPrefix);

String? languageLabel(String tag) {
  for (final entry in languageTags) {
    if (entry.value == tag) {
      return entry.label;
    }
  }
  return null;
}

String? getLanguageTag(List<String> tags) {
  for (final tag in tags) {
    if (isLanguageTag(tag)) {
      return tag;
    }
  }
  return null;
}

String libraryFilterLabel({
  required String keyFilter,
  required String languageFilter,
  required String artistFilter,
}) {
  final parts = <String>[];
  if (keyFilter.isNotEmpty) {
    parts.add(keyFilter);
  }
  if (languageFilter.isNotEmpty) {
    final label = languageLabel(languageFilter);
    if (label != null) {
      parts.add(label);
    }
  }
  if (artistFilter.isNotEmpty) {
    parts.add(artistFilter);
  }
  if (parts.isEmpty) {
    return 'Filter';
  }
  return 'Filter · ${parts.join(' · ')}';
}
